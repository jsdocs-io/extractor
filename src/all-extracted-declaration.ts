import type {
  ExtractedClass,
  ExtractedClassConstructor,
  ExtractedClassMethod,
  ExtractedClassProperty,
} from "./extract-class";
import type { ExtractedEnum, ExtractedEnumMember } from "./extract-enum";
import type { ExtractedFunction } from "./extract-function";
import type {
  ExtractedInterface,
  ExtractedInterfaceCallSignature,
  ExtractedInterfaceConstructSignature,
  ExtractedInterfaceGetAccessor,
  ExtractedInterfaceIndexSignature,
  ExtractedInterfaceMethod,
  ExtractedInterfaceProperty,
  ExtractedInterfaceSetAccessor,
} from "./extract-interface";
import type { ExtractedNamespace } from "./extract-namespace";
import type { ExtractedTypeAlias } from "./extract-type-alias";
import type { ExtractedVariable } from "./extract-variable";

/**
`AllExtractedDeclaration` is the union of all possible declarations
that can be extracted, with some being found only in other declarations
(e.g., class method declarations are found only in a class declaration).
*/
export type AllExtractedDeclaration =
  | ExtractedVariable
  | ExtractedFunction
  | ExtractedClass
  | ExtractedClassConstructor
  | ExtractedClassProperty
  | ExtractedClassMethod
  | ExtractedInterface
  | ExtractedInterfaceProperty
  | ExtractedInterfaceMethod
  | ExtractedInterfaceConstructSignature
  | ExtractedInterfaceCallSignature
  | ExtractedInterfaceIndexSignature
  | ExtractedInterfaceGetAccessor
  | ExtractedInterfaceSetAccessor
  | ExtractedEnum
  | ExtractedEnumMember
  | ExtractedTypeAlias
  | ExtractedNamespace;

/**
`AllExtractedDeclarationKind` is the union of all discriminators
used to detect the kind of declaration.
*/
export type AllExtractedDeclarationKind = AllExtractedDeclaration["kind"];
