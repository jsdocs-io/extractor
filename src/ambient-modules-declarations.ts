import { ModuleDeclaration, Node, type Project } from "ts-morph";
import { isHidden } from "./is-hidden";
import { sourceFilePath } from "./source-file-path";

export type AmbientModulesDeclarationsReturn = {
  containerName: string;
  exportName: string;
  declaration: ModuleDeclaration;
}[];

export const ambientModulesDeclarations = (
  containerName: string,
  project: Project,
  pkgName?: string,
): AmbientModulesDeclarationsReturn => {
  const ambientModulesDeclarations = [];
  for (const symbol of project.getAmbientModules()) {
    for (const declaration of symbol.getDeclarations()) {
      if (isHidden(declaration) || !Node.isModuleDeclaration(declaration)) {
        continue;
      }
      if (pkgName && !sourceFilePath(declaration).startsWith(`/${pkgName}`)) {
        // Ignore ambient modules that are not from the analyzed package.
        continue;
      }
      const exportName = declaration.getName();
      // TODO:
      // Remove surrounding quotes and eventual spaces
      // const exportID = exportName
      //   .replace(/"|'/g, "")
      //   .replace(/\s/g, "_")
      //   .trim();
      ambientModulesDeclarations.push({
        containerName,
        exportName,
        declaration,
      });
    }
  }
  return ambientModulesDeclarations;
};
