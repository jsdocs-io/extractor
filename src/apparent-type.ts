import { TypeFormatFlags, ts, type Node } from "ts-morph";

export function apparentType(node: Node): string {
	// See:
	// https://github.com/dsherret/ts-morph/issues/453#issuecomment-427405736
	// https://github.com/dsherret/ts-morph/issues/453#issuecomment-667578386
	return node
		.getType()
		.getApparentType()
		.getText(
			node,
			ts.TypeFormatFlags.NoTruncation | TypeFormatFlags.UseAliasDefinedOutsideCurrentScope,
		)
		.replace(/^Number$/, "number")
		.replace(/^Boolean$/, "boolean")
		.replace(/^String$/, "string");
}
