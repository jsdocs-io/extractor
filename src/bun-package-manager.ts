import { Effect } from "effect";
import { execa } from "execa";
import { InstallPackageError, PackageManager } from "./package-manager";

/** @internal */
export const bunPackageManager = (bunPath = "bun") =>
	PackageManager.of({
		installPackage: ({ pkg, cwd }) =>
			Effect.gen(function* (_) {
				// Run `bun add <pkg> --verbose`.
				// See https://bun.sh/docs/cli/add.
				const { stdout } = yield* _(
					Effect.tryPromise({
						try: () => execa(bunPath, ["add", pkg, "--verbose"], { cwd }),
						catch: (e) => new InstallPackageError({ cause: e }),
					}),
				);

				// With verbose output on, bun prints one line per installed package
				// (e.g., "foo@1.0.0"), including all installed dependencies.
				// These lines are between the two delimiting lines found here:
				// https://github.com/oven-sh/bun/blob/972a7b7080bd3066b54dcb43e9c91c5dfa26a69c/src/install/lockfile.zig#L5369-L5370.
				const lines = stdout.split("\n");
				const beginHash = lines.findIndex((line) => line.startsWith("-- BEGIN SHA512/256"));
				const endHash = lines.findIndex((line) => line.startsWith("-- END HASH"));
				const installedPackages = lines.slice(beginHash + 1, endHash);
				return installedPackages;
			}),
	});
