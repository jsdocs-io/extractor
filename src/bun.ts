import { execa } from "execa";

/** `Bun` is a wrapper for the Bun package manager. */
export class Bun {
	#bunCmd: string;

	constructor(bunCmd = "bun") {
		this.#bunCmd = bunCmd.trim();
	}

	/**
	`add` installs the given package in the given directory and
	returns a list of all installed packages (e.g., `["foo@1.0.0"]`).

	@see {@link https://bun.sh/docs/cli/add}
	*/
	async add(pkg: string, cwd: string): Promise<string[]> {
		// Run `bun add <pkg> --verbose` in the `cwd`.
		const { stdout } = await execa(this.#bunCmd, ["add", pkg, "--verbose"], { cwd });

		// With verbose output on, bun prints one line per installed package
		// (e.g., "foo@1.0.0"), including all installed dependencies.
		// These lines are between the two delimiting lines found here:
		// https://github.com/oven-sh/bun/blob/972a7b7080bd3066b54dcb43e9c91c5dfa26a69c/src/install/lockfile.zig#L5369-L5370.
		const lines = stdout.split("\n");
		const beginHash = lines.findIndex((line) => line.startsWith("-- BEGIN SHA512/256"));
		const endHash = lines.findIndex((line) => line.startsWith("-- END HASH"));
		const installedPackages = lines.slice(beginHash + 1, endHash);
		return installedPackages;
	}
}
