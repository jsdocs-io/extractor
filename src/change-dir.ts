import { Result } from "neverthrow";
import { OsError } from "./errors";

export const changeDir = Result.fromThrowable(
  process.chdir,
  (e) => new OsError("failed to change directory", { cause: e }),
);
