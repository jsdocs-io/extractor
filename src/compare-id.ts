export type WithId = {
  id: string;
};

export const compareId = (a: WithId, b: WithId): number =>
  a.id.localeCompare(b.id);
