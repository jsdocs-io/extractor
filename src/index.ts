export {
  extractDeclarations,
  type ExtractDeclarationsOptions,
  type ExtractedDeclaration,
  type ExtractedDeclarationKind,
} from "./extract-declarations";
