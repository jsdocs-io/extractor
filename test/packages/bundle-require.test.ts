import { expect, test } from "vitest";
import { extractPackageApi } from "../../src";

test("bundle-require@4.0.2", async () => {
  expect(
    await extractPackageApi({
      pkg: "bundle-require@4.0.2",
    }),
  ).toMatchSnapshot({
    analyzedAt: expect.any(String),
    analyzedIn: expect.any(Number),
    packages: expect.any(Array),
  });
});
