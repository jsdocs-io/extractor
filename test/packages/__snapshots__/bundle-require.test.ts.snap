// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bundle-require@4.0.2 1`] = `
{
  "analyzedAt": Any<String>,
  "analyzedIn": Any<Number>,
  "declarations": [
    {
      "docs": [],
      "file": "/bundle-require/dist/index.d.ts",
      "id": "function.bundleRequire",
      "kind": "function",
      "line": 83,
      "name": "bundleRequire",
      "signature": "bundleRequire: <T = any>(
  options: Options,
) => Promise<{ mod: T; dependencies: string[] }>;",
    },
    {
      "docs": [
        "/**
 * An esbuild plugin to mark node_modules as external
 */",
      ],
      "file": "/bundle-require/dist/index.d.ts",
      "id": "function.externalPlugin",
      "kind": "function",
      "line": 78,
      "name": "externalPlugin",
      "signature": "externalPlugin: ({
  external,
  notExternal,
}?: {
  external?: (string | RegExp)[] | undefined;
  notExternal?: (string | RegExp)[] | undefined;
}) => Plugin;",
    },
    {
      "docs": [],
      "file": "/bundle-require/dist/index.d.ts",
      "id": "function.injectFileScopePlugin",
      "kind": "function",
      "line": 82,
      "name": "injectFileScopePlugin",
      "signature": "injectFileScopePlugin: () => Plugin;",
    },
    {
      "docs": [],
      "file": "/bundle-require/dist/index.d.ts",
      "id": "function.match",
      "kind": "function",
      "line": 74,
      "name": "match",
      "signature": "match: (id: string, patterns?: (string | RegExp)[]) => boolean;",
    },
    {
      "docs": [],
      "file": "/bundle-require/dist/index.d.ts",
      "id": "function.tsconfigPathsToRegExp",
      "kind": "function",
      "line": 73,
      "name": "tsconfigPathsToRegExp",
      "signature": "tsconfigPathsToRegExp: (paths: Record<string, any>) => RegExp[];",
    },
    {
      "callSignatures": [],
      "constructSignatures": [],
      "docs": [],
      "file": "/bundle-require/dist/index.d.ts",
      "getAccessors": [],
      "id": "interface.Options",
      "indexSignatures": [],
      "kind": "interface",
      "line": 22,
      "methods": [],
      "name": "Options",
      "properties": [
        {
          "docs": [],
          "file": "/bundle-require/dist/index.d.ts",
          "id": "interface.Options.property.cwd",
          "kind": "interface-property",
          "line": 23,
          "name": "cwd",
          "signature": "cwd?: string;",
        },
        {
          "docs": [
            "/**
     * esbuild options
     *
     * @deprecated \`esbuildOptions.watch\` is deprecated, use \`onRebuild\` instead
     */",
          ],
          "file": "/bundle-require/dist/index.d.ts",
          "id": "interface.Options.property.esbuildOptions",
          "kind": "interface-property",
          "line": 39,
          "name": "esbuildOptions",
          "signature": "esbuildOptions?: BuildOptions & {
  watch?:
    | boolean
    | {
        onRebuild?: RebuildCallback;
      };
};",
        },
        {
          "docs": [
            "/** External packages */",
          ],
          "file": "/bundle-require/dist/index.d.ts",
          "id": "interface.Options.property.external",
          "kind": "interface-property",
          "line": 58,
          "name": "external",
          "signature": "external?: (string | RegExp)[];",
        },
        {
          "docs": [
            "/**
     * The filepath to bundle and require
     */",
          ],
          "file": "/bundle-require/dist/index.d.ts",
          "id": "interface.Options.property.filepath",
          "kind": "interface-property",
          "line": 27,
          "name": "filepath",
          "signature": "filepath: string;",
        },
        {
          "docs": [
            "/**
     * Provide bundle format explicitly
     * to skip the default format inference
     */",
          ],
          "file": "/bundle-require/dist/index.d.ts",
          "id": "interface.Options.property.format",
          "kind": "interface-property",
          "line": 70,
          "name": "format",
          "signature": "format?: "cjs" | "esm";",
        },
        {
          "docs": [
            "/**
     * Get the path to the output file
     * By default we simply replace the extension with \`.bundled_{randomId}.js\`
     */",
          ],
          "file": "/bundle-require/dist/index.d.ts",
          "id": "interface.Options.property.getOutputFile",
          "kind": "interface-property",
          "line": 48,
          "name": "getOutputFile",
          "signature": "getOutputFile?: GetOutputFile;",
        },
        {
          "docs": [
            "/**
     * Enable watching and call the callback after each rebuild
     */",
          ],
          "file": "/bundle-require/dist/index.d.ts",
          "id": "interface.Options.property.onRebuild",
          "kind": "interface-property",
          "line": 52,
          "name": "onRebuild",
          "signature": "onRebuild?: (ctx: {
  err?: Pick<BuildFailure, "errors" | "warnings">;
  mod?: any;
  dependencies?: string[];
}) => void;",
        },
        {
          "docs": [
            "/**
     * Preserve compiled temporary file for debugging
     * Default to \`process.env.BUNDLE_REQUIRE_PRESERVE\`
     */",
          ],
          "file": "/bundle-require/dist/index.d.ts",
          "id": "interface.Options.property.preserveTemporaryFile",
          "kind": "interface-property",
          "line": 65,
          "name": "preserveTemporaryFile",
          "signature": "preserveTemporaryFile?: boolean;",
        },
        {
          "docs": [
            "/**
     * The \`require\` function that is used to load the output file
     * Default to the global \`require\` function
     * This function can be asynchronous, i.e. returns a Promise
     */",
          ],
          "file": "/bundle-require/dist/index.d.ts",
          "id": "interface.Options.property.require",
          "kind": "interface-property",
          "line": 33,
          "name": "require",
          "signature": "require?: RequireFunction;",
        },
        {
          "docs": [
            "/** A custom tsconfig path to read \`paths\` option */",
          ],
          "file": "/bundle-require/dist/index.d.ts",
          "id": "interface.Options.property.tsconfig",
          "kind": "interface-property",
          "line": 60,
          "name": "tsconfig",
          "signature": "tsconfig?: string;",
        },
      ],
      "setAccessors": [],
      "signature": "interface Options {}",
    },
    {
      "docs": [],
      "file": "/bundle-require/dist/index.d.ts",
      "id": "type.GetOutputFile",
      "kind": "type",
      "line": 20,
      "name": "GetOutputFile",
      "signature": "type GetOutputFile = (filepath: string, format: "esm" | "cjs") => string;",
    },
    {
      "docs": [],
      "file": "/bundle-require/dist/index.d.ts",
      "id": "type.RebuildCallback",
      "kind": "type",
      "line": 21,
      "name": "RebuildCallback",
      "signature": "type RebuildCallback = (
  error: Pick<BuildFailure, "errors" | "warnings"> | null,
  result: BuildResult | null,
) => void;",
    },
    {
      "docs": [],
      "file": "/bundle-require/dist/index.d.ts",
      "id": "type.RequireFunction",
      "kind": "type",
      "line": 17,
      "name": "RequireFunction",
      "signature": "type RequireFunction = (
  outfile: string,
  ctx: {
    format: "cjs" | "esm";
  },
) => any;",
    },
    {
      "docs": [
        "/**
 * Dynamically import files.
 *
 * As a temporary workaround for Jest's lack of stable ESM support, we fallback to require
 * if we're in a Jest environment.
 * See https://github.com/vitejs/vite/pull/5197#issuecomment-938054077
 *
 * @param file File path to import.
 */",
      ],
      "file": "/bundle-require/dist/index.d.ts",
      "id": "variable.dynamicImport",
      "kind": "variable",
      "line": 13,
      "name": "dynamicImport",
      "signature": "const dynamicImport: RequireFunction;",
    },
    {
      "docs": [],
      "file": "/bundle-require/dist/index.d.ts",
      "id": "variable.JS_EXT_RE",
      "kind": "variable",
      "line": 15,
      "name": "JS_EXT_RE",
      "signature": "const JS_EXT_RE: RegExp;",
    },
  ],
  "name": "bundle-require",
  "overview": undefined,
  "packages": Any<Array>,
  "subpath": ".",
  "types": "./dist/index.d.ts",
  "version": "4.0.2",
}
`;
